Support 24h time formats

josefarias opened the pull request:
This PR adds support for 24-hour time formats.

This is how it works:

First, configure the library to use 24-hour time formats:
```javascript
import LocalTime from "local-time"

LocalTime.config.useFormat24 = true
```

Then, render a `<time>` element with both 12 and 24 hour formats.
Use `data-format24` to specify the 24h format and the usual `data-format` to specify the 12h format.
```html
<time data-format="%-l:%M%P" data-format24="%H:%M" data-local="time" datetime="2023-06-26T06:00:00.000Z">6:00am</time>
```

The rails helpers have been modified to auto-fill `data-format24` when using named formats. They will look for equivalent keys with the `_24h` suffix in i18n and date & time configurations.

By default, the library will use `data-format`. But it will instead use `data-format24` if `config.format24` is `true` at runtime. The library will also fall back to `data-format` if `config.format24` is set to `true` but no `data-format24` value is provided.

This does not change existing behavior as this feature is purely opt-in via `config.useFormat24`.

josefarias commented on test/helpers/local_time_helper_test.rb:
This file used to refer to a format defined in `Time::DATE_FORMATS[format]` and/or `Date::DATE_FORMATS[format]` as a `date_formats_format`. I thought that was difficult to grok because of the repetition in the name and the fact that we're not always dealing with _date_ formats. So I changed it to `ruby_format` to communicate that the format is defined in ruby constants.

ramhoj commented on app/helpers/local_time_helper.rb:
```diff
@@ -67,4 +53,33 @@ def extract_options_and_value(options, value_key = nil)
+      end
+    end
+
+    def find_time_format_by_name(format)
```
What do you think about naming this `find_time_formats_by_name` to make it more obvious that it nows returns two formats?

ramhoj commented on app/helpers/local_time_helper.rb:
```diff
@@ -67,4 +53,33 @@ def extract_options_and_value(options, value_key = nil)
+        format12, format24 = nil, nil
+      end
+
+      [format12.presence || LocalTime.default_time_format, format24.presence]
```
Would it be possible to move the defaults to `i18n_time_or_date_format` and `ruby_time_or_date_format` and to make this method a bit easier to read? If so we could remove the `||` here and the `else` above.

josefarias replied:
So `i18n_time_or_date_format` and `ruby_time_or_date_format` are actually serving a dual purpose. They return the format, but they're also specifying whether we should use an `i18n` or `ruby` strategy for looking up the `_24h` counterpart.

Moving the defaults there would interfere with that logic. Which simply indicates this _does_ need a rewrite.

Clean this up in https://github.com/basecamp/local_time/pull/135/commits/4aba3fb1e0f08d6293623cb717161595d10e32b5, let me know what you think.

ramhoj replied:
This reads much better now, thank you!

ramhoj commented on app/helpers/local_time_helper.rb:
```diff
@@ -67,4 +53,33 @@ def extract_options_and_value(options, value_key = nil)
+      if format12 = i18n_time_or_date_format(format)
+        format24 = i18n_time_or_date_format("#{format}_24h")
+      elsif fmt = ruby_time_or_date_format(format)
+        format12 = fmt.is_a?(Proc) ? nil : fmt
```
Do we need this? I can only see that we're calling this from [here](https://github.com/basecamp/local_time/pull/135/files#diff-d0283343941c298e4f4eeebb501cfb6b01b2380b927ea6e14a7c7a9f771528f6R59) where we've already guarded on sym?

ramhoj reviewed (commented):
Thank you @josefarias! I added a few minor comments, feel free to dismiss where I might be uninformed.

jorgemanrubia reviewed (commented):
This looks fantastic @josefarias 👏

Some good suggestions from @ramhoj, and a couple of notes from me, but all super minor and just for your consideration.

josefarias commented on lib/assets/javascripts/src/local-time/controller.coffee:
```diff
@@ -27,8 +27,11 @@ class LocalTime.Controller
     datetime = element.getAttribute("datetime")
-    format = element.getAttribute("data-format")
     local = element.getAttribute("data-local")
+    format = if config.format24 and element.hasAttribute("data-format24")
```
> as we're coming from 12h only

Yeah I think this is what breaks it for me. I'd love to have symmetry with ISO and Intl.DateTimeFormat (https://github.com/basecamp/local_time/pull/135#discussion_r1236446931) but basing the conditionals and config off of `hour12` would depart from the existing assumption that this is a 12-hour-first library. If only because it was built that way initially.

This assumption is present, for example, in the convention that `data-format` should be used for 12-hour formats, whereas `data-format24` should be used for 24-hour formats.

In my mind, `data-format` is clearly the norm and `data-formatXX` departs from that norm. If hour12 is the norm, then checking if we're following the norm is strange. I'd rather check if we're departing from the norm, if that makes sense.

ramhoj replied:
Thank you for the context @josefarias. Given this I agree that this change is too big here, good call!

josefarias commented on lib/assets/javascripts/src/local-time/config/format24.coffee:
```diff
@@ -0,0 +1,3 @@
+import LocalTime from "../local_time"
+
+LocalTime.config.format24 = false
```
Addressed this in https://github.com/basecamp/local_time/pull/135#discussion_r1237255204 let me know what you think!

I thought about using `hour24` instead. But I like the symmetry `config.format24` has with `data-format24`. It allows to universally refer to this feature as just `format24` in filenames, configs, APIs, etc.

ramhoj replied:
Nice, I like that!

josefarias commented on app/helpers/local_time_helper.rb:
```diff
@@ -67,4 +53,33 @@ def extract_options_and_value(options, value_key = nil)
+      if format.is_a?(Symbol)
+        find_time_format_by_name(format)
+      else
+        [format.presence || LocalTime.default_time_format, nil]
```
Ah, thanks! I'll fix this formatting here and add rubocop in a separate PR to avoid polluting the diffs here. I checked and rubocop would touch more than a couple files — all with minor, changes.

josefarias commented on test/javascripts/src/format24_test.coffee:
```diff
@@ -0,0 +1,46 @@
+
+    el = addTimeEl format: "%-l:%M%P", format24: "%H:%M", datetime: now.toISOString()
+    defer ->
+      assert.equal getText(el), now.format("H:mm")
```
Great catch!

josefarias commented on README.md:
Not using travis-ci or saucelabs anymore

josefarias commented:
Thanks for the reviews @ramhoj, @jorgemanrubia!

Addressed the feedback, made some comments, and added a readme entry explaining how the feature works.

This is ready for another round of reviews.

jorgemanrubia reviewed (approved):
👏

ramhoj reviewed (commented):
Great work @josefarias, looks great!
